support argmax/argmin without dim kwargs and fix adaptive_max_pool3d

cccclai opened the pull request:
Summary: As title, in PyTorch, when dim is not set, it will flatten the input and get argmax as dim=0. Add a pass to reshape the input when dim is not set and consolidate test case

edit:
1. Apply to argmin too
2. Add `exir_ops.edge.aten.adaptive_max_pool3d.default` to the to be implemented op list to pass the error

Differential Revision: D83606497

haowhsu-quic commented on backends/qualcomm/_passes/insert_reshape_for_argmax.py:
```diff
@@ -0,0 +1,59 @@
+        graph = graph_module.graph
+        modified = False
+
+        for n in list(graph.nodes):
```
I think we can use the iterator directly: `for n in graph.nodes:`.

cccclai replied:
updated

haowhsu-quic commented on backends/qualcomm/_passes/insert_reshape_for_argmax.py:
```diff
@@ -0,0 +1,59 @@
+                    with graph.inserting_before(n):
+                        reshape_node = graph.create_node(
+                            "call_function",
+                            torch.ops.aten.reshape.default,
```
I think we use `torch` namespace for graph which needs decomposition before `prepare_pt2e` and `exir_ops` namespace for floating point graph which goes directly to `to_edge_transform_and_lower` API.

haowhsu-quic replied:
It's a convention we're using but I guess the op builder could still work since it only recognizes part of op's target string.

cccclai replied:
I was actually using it for both `transform_for_annotation_pipeline` and `transform_for_export_pipeline`. I was thinking it could be a good idea to have it in preprocess, but preprocess happens after partition, and op_argmax.py will run during partition as well. I think in this case, we need to have it similar to those decompose passes, which means we use `torch` namespace

haowhsu-quic commented on backends/qualcomm/_passes/insert_reshape_for_argmax.py:
```diff
@@ -0,0 +1,59 @@
+                    # Rewrite argmax: take reshape_node as input, set dim=0
+                    n.args = (reshape_node, 0, *n.args[2:])
+                
+                modified = True
```
I think you need one more indention to have this statement fall into branch.

cccclai replied:
updated

haowhsu-quic commented on backends/qualcomm/tests/test_passes.py:
```diff
@@ -0,0 +1,42 @@
+import torch
+from executorch.backends.qualcomm._passes import InsertReshapeForArgmax
+
+class TestPasses(unittest.TestCase):
```
Thank you for adding this file, it's helpful for us to test all the passes thoroughly.

mergennachin commented on backends/qualcomm/builders/op_argmax.py:
```diff
@@ -46,7 +46,7 @@ def define_node(
         )
         argmax_output_tensors = [argmax_out_tensor_wrapper]
-
+        print("argmax node.args: ", node.args)
```
remove

cccclai replied:
updated

cccclai commented on backends/qualcomm/_passes/insert_reshape_for_reduce_ops.py:
```diff
@@ -0,0 +1,59 @@
+# Copyright (c) Qualcomm Innovation Center, Inc.
```
updated

cccclai commented:
Can I get another round of review on this? I fix a few more failing tests including `argmin`, and `adaptive_max_pool3d`

haowhsu-quic reviewed (approved):
Thank you!

cccclai commented on backends/qualcomm/partition/common_defs.py:
```diff
@@ -17,6 +17,7 @@
 to_be_implemented_operator = [
     exir_ops.edge.aten._adaptive_avg_pool3d.default,
     exir_ops.edge.aten.adaptive_max_pool2d.default,
+    exir_ops.edge.aten.adaptive_max_pool3d.default,
```
Yeah updated

cccclai committed:
support argmax/argmin without dim kwargs and fix adaptive_max_pool3d (#14710)

Summary:

As title, in PyTorch, when dim is not set, it will flatten the input and get argmax as dim=0. Add a pass to reshape the input when dim is not set and consolidate test case

Differential Revision: D83606497

cccclai commented:
@pytorchbot cherry-pick --onto release/1.0 -c regression
